Source/WebCore/html/HTMLElement.cpp

```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
     return { };
 }
 
-static HTMLElement* topmostPopoverAncestor(Element& element)
-{
-    for (auto& element : lineageOfType<HTMLElement>(element)) {
-        if (element.popoverState() != PopoverState::Auto)
-            continue;
-        if (element.popoverData()->visibilityState() != PopoverVisibilityState::Showing)
+// https://html.spec.whatwg.org/#topmost-popover-ancestor
+// Consider both DOM ancestors and popovers where the given popover was invoked from as ancestors.
+// Use top layer positions to disambiguate the topmost one when both exist.
+static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+{
+    // Store positions to avoid having to do O(n) search for every popover invoker.
+    HashMap<Ref<Element>, size_t> topLayerPositions;
+    size_t i = 0;
+    for (auto& element : newPopover.document().topLayerElements()) {
+        if (!is<HTMLElement>(element) || downcast<HTMLElement>(element.get()).popoverState() != PopoverState::Auto)
             continue;
-        return &element;
+        topLayerPositions.add(element, i);
+        i++;
     }
 
-    // FIXME: Include popover invokers and compare them in top layer order once they are implemented.
+    auto nearestInclusiveOpenPopover = [](Element& candidate) -> HTMLElement* {
+        for (auto& element : lineageOfType<HTMLElement>(candidate)) {
+            if (element.popoverState() == PopoverState::Auto && element.popoverData()->visibilityState() == PopoverVisibilityState::Showing)
+                return &element;
+        }
+        return nullptr;
+    };
 
-    return nullptr;
+    RefPtr topmostAncestor = nearestInclusiveOpenPopover(newPopover);
+
+    // Iterate over all popover invokers in the document.
```

Comment:
The spec says that popover invokers are buttons. However, I don't see the logic in the loop below. We are iterating over all HTMLFormControlElement, which is not quite the same.